design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,163 @@
+# Overview
+This proposal is about adding CPU hotplug support to KubeVirt.
+CPU hotplug will allow users to dynamically add/remove CPU resources to/from running VMs.
+While this design is focused on CPU hotplug, it aims to provide a generic API design for resource hotplug that will be reusable for things like memory.
+
+## Motivation
+CPU hotplug is now a common virtualization feature, but KubeVirt lacks support for it.
+Live vertical scaling of VM resources in general can be very useful in a variety of scenarios.
+
+## Goals
+- Hotplug and hot-unplug of CPU resources from VMs should be achievable by editing the VM object itself (no sub-resources)
+- Underlying technology used for hotplug should be independent of the VM API, meaning we should be able to swap out how hotplug is technically performed in the future without impacting API compatibility
+- Implementation should be achievable today, with Kubernetes APIs that are at least in beta. Unfortunately, at the time of writing, the Kubernetes vertical pod scaling API is only alpha
+
+## Non Goals
+- Revising existing hotplug mechanisms like Volume hotplug
+- Dynamically applying affinity rules to VMIs
+
+## Definition of Users
+A user with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+## User Stories
+As a VM user, I would like to dynamically increase or decrease the number of CPUs used by a VM without requiring the VM to restart.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+In VM objects, the existing CPU `sockets` field will now be dynamic, provided that it has been declared as such.
+Declaring the field as dynamic will be done by adding a `cpu` entry to a new section under the VM spec, called `liveUpdateFeatures`.
+Under the `cpu` entry, it will be possible to define a maximum number of sockets. That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set. That default value will be configurable in the KubeVirt CR.
```

Comment:
Is this field mutable?

Reply:
Everything in kubevirt CR is mutable, no?

Reply:
I thought this refers mainly to `VM.spec.liveUpdateFeatures.cpu`. KV is of course mutable and I don't see a problem there.

Reply:
Right, changing the number of maxCpus in the `liveUpdateFeatures.cpu` struct shouldn't have any effect on the running VMI. It should only propagate to the vmi spec on re-creation.